core/App/ehsm_napi.cpp

```diff
@@ -84,35 +83,50 @@ void NAPI_Finalize(){
     Finalize();
 }
 
-/*
-@return
-[string] json string
+/**
+ * @brief Create key and save the parameters when using the key for encrypt, decrypt, sign and verify
+ * 
+ * @param paramJson Pass in the key parameter in the form of JSON string
+ * @return char* 
+ * [string] json string
     {
         code: int,
         message: string,
         result: {
-            cmk_base64 : string,
+            cmk_base64 : string
         }
     }
-*/
-char* NAPI_CreateKey(const uint32_t keyspec, const uint32_t origin)
+ */
+char* NAPI_CreateKey(const char* paramJson)
 {
     RetJsonObj retJsonObj;
     ehsm_status_t ret = EH_OK;
     ehsm_keyblob_t master_key;
-
-    memset(&master_key, 0, sizeof(master_key));
-
     string cmk_base64;
-
     uint8_t *resp = NULL;
     uint32_t resp_len = 0;
 
-    master_key.metadata.keyspec = keyspec;
-    master_key.metadata.origin = origin;
-    master_key.keybloblen = 0;    
+    // parse paramJson into paramJsonObj
+    JsonObj paramJsonObj;
+    if (!paramJsonObj.parse(paramJson))
+    {
+        retJsonObj.setCode(retJsonObj.CODE_FAILED);
+        retJsonObj.setMessage("Server exception.");
+        goto out;
+    }
+
+    memset(&master_key, 0, sizeof(master_key));
 
-    ret = CreateKey(&master_key);
+    
+
+    // storage common key properties into metadata of master_key 
+    master_key.metadata.keyspec = paramJsonObj.readData_uint16("keyspec");
+    master_key.metadata.origin = paramJsonObj.readData_uint16("origin");
+    master_key.metadata.purpose = paramJsonObj.readData_uint16("purpose");
+    master_key.keybloblen = 0;    
+    
+    // get keybloblen
+    ret = CreateKey(&master_key, paramJsonObj);
```

Comment:
what does the paramJsonObj use for? i don't think you need to pass it into low level API.

Reply:
yes, and I'll remove it